Code/GraphMol/SubstanceGroup.cpp

```diff
@@ -294,6 +295,74 @@ unsigned int addSubstanceGroup(ROMol &mol, SubstanceGroup sgroup) {
   return id;
 }
 
+namespace {
+bool includesBond(SubstanceGroup &sg, unsigned int idx) {
+  return sg.includesBond(idx);
+}
+bool includesAtom(SubstanceGroup &sg, unsigned int idx) {
+  return sg.includesAtom(idx);
+}
+void removedBond(SubstanceGroup &sg, unsigned int idx) {
+  sg.adjustToRemovedBond(idx);
+}
+void removedAtom(SubstanceGroup &sg, unsigned int idx) {
+  sg.adjustToRemovedAtom(idx);
+}
+
+template <bool INCLUDES_METHOD(SubstanceGroup &, unsigned int),
+          void ADJUST_METHOD(SubstanceGroup &, unsigned int)>
+void removeSubstanceGroupsReferencing(RWMol &mol, unsigned int idx) {
+  auto &sgs = getSubstanceGroups(mol);
+  if (!sgs.empty()) {
+    // first collect a vector the ones that should be removed
+    std::vector<unsigned> toRemove;
+    for (auto &&sg : sgs) {
+      if (INCLUDES_METHOD(sg, idx)) {
+        unsigned int index;
+        if (sg.getPropIfPresent("index", index)) {
+          toRemove.push_back(index);
+        }
+      }
+    }
+    // now go through and keep everything that shouldn't be removed
+    // and who doesn't have a PARENT equal to one that should be removed.
+    std::vector<SubstanceGroup> newsgs;
+    newsgs.reserve(sgs.size());
+    for (auto &&sg : sgs) {
+      unsigned int index;
+      if (sg.getPropIfPresent("index", index)) {
+        unsigned int parent = 0xf00d;
+        if (std::find(toRemove.begin(), toRemove.end(), index) ==
+                toRemove.end() &&
+            (!sg.getPropIfPresent("PARENT", parent) ||
+             std::find(toRemove.begin(), toRemove.end(), parent) ==
+                 toRemove.end())) {
+          ADJUST_METHOD(sg, idx);
+          newsgs.push_back(std::move(sg));
+        }
+      } else if (!INCLUDES_METHOD(sg, idx)) {
```

Comment:
Hum, this doesn't capture multiple levels of parenting, does it? (I know, it is a rare possibility, but it is possible).

Reply:
good point. I forgot about that. It's fixed now